### ISSUE TITLE
YOLOv8 Onnx Deploy

### ISSUE THREAD

**Issue by ProgrammerZhujinming** [ultralytics/ultralytics#14962]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi! I trained the yolov8 model with my custom dataset. I export the model as onnx format and it only accept input with shape(640*640), but when I use the model to interface my test image which has the resolution about 1944 * 2592, I found it will resize my image to 480 * 640 and object can be detected. But the onnx weight can only accept the input with 640 * 640. I read the code about preprocess but I cannot find some code to handle this situation.   
`
    stride = 32

    ori_img = cv2.imread("datasets/yimu/images/train/11.bmp")
    h, w, c = ori_img.shape

    shape = ori_img.shape[:2]  # current shape [height, width]
    new_shape = [640, 640]

    r = min(new_shape[0] / shape[0], new_shape[1] / shape[1])

    # Compute padding
    ratio = r, r  # width, height ratios
    new_unpad = int(round(shape[1] * r)), int(round(shape[0] * r))
    dw, dh = new_shape[1] - new_unpad[0], new_shape[0] - new_unpad[1]  # wh padding

    dw, dh = np.mod(dw, stride), np.mod(dh, stride)  # wh padding

    dw /= 2  # divide padding into 2 sides
    dh /= 2

    if shape[::-1] != new_unpad:  # resize
        ori_img = cv2.resize(ori_img, new_unpad, interpolation=cv2.INTER_LINEAR)

    #top, bottom = int(round(dh - 0.1))
    #left, right = int(round(dw - 0.1))

    top, bottom = 0, int(round(dh + 0.1))
    left, right = 0, int(round(dw + 0.1))

    ori_img = cv2.copyMakeBorder(
        ori_img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=(114, 114, 114)
    )  # add border

    ori_img = np.expand_dims(ori_img, axis=0)

    ori_img = ori_img[..., ::-1].transpose((0, 3, 1, 2))  # BGR to RGB, BHWC to BCHW, (n, 3, h, w)
    ori_img = np.ascontiguousarray(ori_img)  # contiguous
    img_tensor = torch.from_numpy(ori_img)

    img_tensor = img_tensor / 255
`
this is important because if I cannot keep consistent with your handling method I cannot get the correct result. 
Actually, I use Triton to deploy the model, so if I need another ways to keep my input consistent with the training process, please offer me related code to solve this problem. Thanks for your reply.

**Comment by Skillnoob** [comment]
If you're using python you can use ultralytics for running inference with your exported onnx model:
```python
from ultralytics import YOLO

model = YOLO('model.onnx')

results = model.predict(source='datasets/yimu/images/train/11.bmp')
```

**Comment by ProgrammerZhujinming** [comment]
Sorry, maybe you don't know my question.
I use triton to deploy yolov8-onnx and I found that onnx only accept input with shape(480, 640) and tensor data format.
However, if I input opencv image format, the model can preprocess the input shape to (480, 640) and infer with it.
Triton need I to fix the input shape and only execute the compute graph, so I got the problem. I don't know if I state my problem clearly and I am sorry to bother you again.

**Comment by pderrenger** [comment]
@ProgrammerZhujinming thank you for clarifying. When deploying YOLOv8 models with Triton, ensure your input preprocessing matches the model's expected input shape and format. The ONNX model expects a fixed input size, typically (640, 640). You need to preprocess your images to this size before passing them to Triton. This involves resizing and padding your images to maintain aspect ratio, then converting them to the required tensor format. If you need further assistance with preprocessing code, please refer to the Ultralytics documentation or the Triton Inference Server documentation for detailed guidance.

**Comment by Y-T-G** [comment]
@ProgrammerZhujinming What `imgsz` argument did you use during training? It's 649 by default.

Why do you want to run it at such a high resolution? 640x640 is default resolution. You don't have to run to run the model at your original image size. 640x640 is enough in most cases.

**Comment by Y-T-G** [comment]
PyTorch models are not resized to 640x640. Only resized to fit the minimum side divisible by 32 and maximum equal to `imgsz`

Set `auto=False` if you want to disable that.

https://github.com/ultralytics/ultralytics/blob/5d9046abda3d1070bc2c4bf234a1ab81f50e43e5/ultralytics/data/augment.py#L1501

**Comment by Jordan-Pierce** [comment]
@Y-T-G I guess I'm just now realizing, but does this mean that if we `model.export(format=torchscript)`, using those weights with `ultralytics` i.e., `model = YOLO(best.torchscript)` isn't necessarily going to give performance boosts, and instead we should be using `torchscript` library instead (or `onnx`, `TenorRT`, ect)? I think I just assumed that `ultralytics` dealt with the inferencing behind the scenes given the model format type. But I'm seeing that you can specify an `imgz` when exporting certain format types, but when using those models with `ultralytics`, it's not using the `imgsz` specified and falls back to 640 (even when trained on something else).

**Comment by Y-T-G** [comment]
@Jordan-Pierce `ultralytics` wouldn't be slower by a lot. It uses the runtime of those formats internally. There could be some overhead due to the abstraction and other quality of life features that `ultralytics` has.

Most export formats will require you to use the `imgsz` that was specified during export unless you exported with `dynamic=True`. If you specify an `imgsz` other than that, it should throw an error during prediction.
